pkg/storage/etcd3: correctly validate resourceVersions

stevekuznetsov opened the pull request:
pkg/storage/etcd3: correctly validate resourceVersions

In a number of tests, the underlying storage backend interaction will
return the revision (logical clock underpinning the MVCC implementation)
at the call-time of the RPC. Previously, the tests validated that this
returned revision was exactly equal to some previously seen revision.
This assertion is only true in systems where no other events are
advancing the logical clock. For instance, when using a single etcd
cluster as a shared fixture for these tests, the assertion is not valid
any longer. By checking that the returned revision is no older than the
previously seen revision, the validation logic is correct in all cases.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

Depends on https://github.com/kubernetes/kubernetes/pull/108936

/kind cleanup

```release-note
NONE
```

```docs

```

/sig api-machinery
/assign @liggitt @smarterclayton @sttts @deads2k

stevekuznetsov committed:
pkg/storage/etcd3: correctly validate resourceVersions

In a number of tests, the underlying storage backend interaction will
return the revision (logical clock underpinning the MVCC implementation)
at the call-time of the RPC. Previously, the tests validated that this
returned revision was exactly equal to some previously seen revision.
This assertion is only true in systems where no other events are
advancing the logical clock. For instance, when using a single etcd
cluster as a shared fixture for these tests, the assertion is not valid
any longer. By checking that the returned revision is no older than the
previously seen revision, the validation logic is correct in all cases.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

liggitt commented:
I'm unsure about this change ... it seems like *something* in the etcd3 store unit tests should be checking that the resourceVersion is stable in the absence of other changes. Isn't the unit test starting its own etcd instance? I'm not sure a shared etcd is a risk here

stevekuznetsov commented:
@liggitt it certainly depends on what the point of the test is. If the test only passes on specific etcd fixtures, it's not really testing that the underling `storage.Interface` is providing the correct mechanisms as necessary for k8s to function. Given that, the test is less precise than it could be and does a bad job of documenting what, specifically, k8s expects from the underlying storage layer.

/retest

stevekuznetsov commented:
> it seems like something in the etcd3 store unit tests should be checking that the resourceVersion is stable in the absence of other changes

I don't think anything in k8s requires that this be the case. It's not an invariant that influences any interaction k8s has with the underlying store to my knowledge. If I'm mistaken here, I'm more than happy to write exhaustive tests to enforce this behavior from etcd.

stevekuznetsov commented:
/retest

fedebongio commented:
/triage accepted

cici37 commented:
/triage accepted

cici37 commented:
/triage accepted

stevekuznetsov commented:
@liggitt @wojtek-t @smarterclayton would love to circle back to this PR. The long-and-short of it is that `etcd` and `etcd` itself decides when the internal revision has changed, and why. Among the actions that get committed to the Raft log and bump the revision are quorum reads, so even in cases where no state has been mutated, the revision may and will increase. Kubernetes exposes the raw current revision as the `resourceVersion` for `List()` calls. The tests today expect the `resourceVersion` on `List()` calls to _exactly match_ a previous version. This is true if and only if the etcd cluster in question has no activities going on (or only non-quorum reads). This is, in my opinion:

 - hyper-specific to the point of being misleading about what, specifically, Kubernetes expects from a conformant etcd cluster
 - fragile in that it fails on just about every etcd fixture except for the particular ones used in these tests

By changing the tests to make the check that I've proposed, we gain readability, document the requirements on etcd better and make the tests less fragile.

smarterclayton commented:
Would `kine` be better testable if this was fixed?  Is there any *real* disadvantage to weakening this assumption (such as compaction / future etcd changes that lead to potentially background changes)?

liggitt commented:
this is in my backlog but I won't be able to dig into it until next week at the earliest; consumed by 1.24 stuff atm

stevekuznetsov commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:
```diff
@@ -1385,7 +1386,7 @@ func TestList(t *testing.T) {
 			expectedRemainingItemCount: utilpointer.Int64Ptr(1),
 			rv:                         "0",
-			expectRV:                   list.ResourceVersion,
+			expectRVFunc:               resourceVersionNoLaterThan(list.ResourceVersion),
```
At a minimum my intent was to be able to run the existing tests in parallel on one `etcd` fixture. As that's actually how Kubernetes uses `etcd`, your option 1 is already the case in every Kubernetes cluster that has concurrent access to the database. Tests that fail in such a scenario are IMHO not super useful in documenting the expectations that Kubernetes has on the underlying storage layer.

liggitt replied:
I haven't even read the changed code in this PR (will have to wait until next week), but following the comment thread, I have a couple reactions:

> Tests that fail in such a scenario are IMHO not super useful in documenting the expectations that Kubernetes has on the underlying storage layer.

For a unit test, I disagree. This is bridging two systems, so testing specific expected behavior of the other system in isolation is exactly what I would expect.

> This test assumption does not apply to any non-toy etcd

I would replace "non-toy" with "non-isolated", but testing in isolation is exactly what I want from a unit test.

stevekuznetsov replied:
@liggitt why is it useful to document expectations of a _single-threaded_ interaction when no part of the system expects the same or makes any use of the behavior which may occur in that manner? `storage.Interface` and `storage/etcd3.store` are not single-threaded libraries and are not used as such.

liggitt replied:
> @liggitt why is it useful to document expectations of a single-threaded interaction when no part of the system expects the same or makes any use of the behavior which may occur in that manner?

To exercise and verify the etcd behavior the etcd3/store.go implementation is based on (revisions bump on writes, not reads, etc).

> maybe the test could be construed as technically useful for a particular isolated etcd fixture but as there are no (!!) tests for storage.Interface, these tests also IMHO serve as documentation for that interface to the rest of the codebase

if we want to separately provide `storage.Interface` tests (and ensure the etcd3 store implementation passes them), that could be reasonable, but I wouldn't weaken the etcd store unit tests in order to do that.

stevekuznetsov replied:
> (revisions bump on writes, not reads, etc).

This is not true today. Unless you specifically ask `etcd` for non-quorum reads, your read bumps the Raft log and the revision.

>  but I wouldn't weaken the etcd store unit tests in order to do that.

Again, this behavior is happenstance for this test fixture and has no implication in any place downstream from etcd. No part of any system - from the `*etcd3.storage` implementation to the end client of the k8s API - ever makes any such assumption.

stevekuznetsov replied:
Also I feel like I'm not communicating this appropriately but the `resourceVersion` we give to clients in the `*List` object is the _current global revision for the Raft members_ and therefore changes with _any_ event that requires consensus in the system. The one and only way that this number does not change is if no client interacts with etcd. Absolutely no part of the current `storage/etcd3.store` implementation makes any assumption whatsoever about this behavior, as it's entirely emergent behavior.

liggitt replied:
> This is not true today. When using linearized reads, your request has to go through Raft consensus, if it's in the Raft log the revision is bumped/

that's not accurate. this sets up a 3-member cluster, does linearized reads across them, and demonstrates the revision does not change:

```shell
IP=$(hostname -I | awk '{print $1}')
CLUSTER=etcd0=http://${IP}:2380,etcd1=http://${IP}:2381,etcd2=http://${IP}:2382
for i in $( seq 0 2 ); do
  docker run --rm -d -p 400${i}:400${i} -p 238${i}:238${i} -p 237${i}:237${i} \
    --hostname etcd${i} \
    --name etcd${i} quay.io/coreos/etcd:v3.4.18 \
    etcd \
    -name etcd${i} \
    -advertise-client-urls http://${IP}:237${i},http://${IP}:400${i} \
    -listen-client-urls http://0.0.0.0:237${i},http://0.0.0.0:400${i} \
    -initial-advertise-peer-urls http://${IP}:238${i} \
    -listen-peer-urls http://0.0.0.0:238${i} \
    -initial-cluster-token etcd-cluster-1 \
    -initial-cluster ${CLUSTER} \
    -initial-cluster-state new
done

# write to first member
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl put /testkey1 testvalue1
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl put /testkey2 testvalue2
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl put /testkey3 testvalue3

# fetch from etcd0
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl get --consistency=l -w json /testkey1
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl get --consistency=l -w json /testkey2
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl get --consistency=l -w json /testkey3
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl get --consistency=l -w json --prefix / | jq .

# fetch from etcd1
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2371 etcdctl get --consistency=l -w json /testkey1
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2371 etcdctl get --consistency=l -w json /testkey2
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2371 etcdctl get --consistency=l -w json /testkey3
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2371 etcdctl get --consistency=l -w json --prefix / | jq .

# fetch from etcd2
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2372 etcdctl get --consistency=l -w json /testkey1
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2372 etcdctl get --consistency=l -w json /testkey2
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2372 etcdctl get --consistency=l -w json /testkey3
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2372 etcdctl get --consistency=l -w json --prefix / | jq .
```

liggitt replied:
> Also I feel like I'm not communicating this appropriately but the resourceVersion we give to clients in the *List object is the current global revision for the Raft members and therefore changes with any event that requires consensus in the system. The one and only way that this number does not change is if no client interacts with etcd. Absolutely no part of the current storage/etcd3.store implementation makes any assumption whatsoever about this behavior, as it's entirely emergent behavior.

Understood. Because the etcd store impl builds things on top of that behavior, that still seems worthwhile to test in isolation.

stevekuznetsov replied:
@liggitt hm - my apologies. All of the official documentation says that the linearized read goes through the full Raft process but it looks like they piggy-back on another message so that the read does not technically bump the log. Slightly misleading documentation.

I don't really understand your point, though - our contract for LIST is that we return the current revision, full stop. Testing that "nothing happens" when "nothing is written" is not really the point here.

stevekuznetsov commented:
@liggitt thanks for the confirmation :)

stevekuznetsov committed:
pkg/storage/etcd3: validate revision invariant

We must ensure that we notice if the etcd behavior on linearized reads
changes.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov commented:
@liggitt added a focused test for the linearized read invariant we spoke about.

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/linearized_read_test.go:
```diff
@@ -0,0 +1,64 @@
+	// ultimately client-facing. By default, all the requests that our *etcd3.store{} issues are linearized.
+	// However, this also includes *read* requests, and we would not expect non-mutating requests against etcd
+	// to, by "go[ing] through the Raft consensus process," result in a higher resource version on List calls.
+	// Today, etcd manages to "go through the Raft consensus process" by piggy-backing on other messages, so
+	// linearized reads do *not* bump the key-value store revision. This test exists to ensure that we notice
```
I think you're confusing "raft index" and "store revision". Those are distinct things. A linearized read that didn't manage to piggy-back on another message *would* bump the raft index but would *not* bump the store version. Only writes bump the store verison.

In the example in https://github.com/kubernetes/kubernetes/pull/108938#discussion_r844024255, you can see this by doing a list and a dump of endpoint status:

```shell
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl get --consistency=l -w json --prefix / | jq .
```

```json
{
  "header": {
    "cluster_id": 16248273405584523000,
    "member_id": 3412179229575531000,
    "revision": 4,
    "raft_term": 2
  },
  "kvs": [
    {
      "key": "L3Rlc3RrZXkx",
      "create_revision": 2,
      "mod_revision": 2,
      "version": 1,
      "value": "dGVzdHZhbHVlMQ=="
    },
    {
      "key": "L3Rlc3RrZXky",
      "create_revision": 3,
      "mod_revision": 3,
      "version": 1,
      "value": "dGVzdHZhbHVlMg=="
    },
    {
      "key": "L3Rlc3RrZXkz",
      "create_revision": 4,
      "mod_revision": 4,
      "version": 1,
      "value": "dGVzdHZhbHVlMw=="
    }
  ],
  "count": 3
}
```

```shell
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370,127.0.0.1:2371,127.0.0.1:2372 etcdctl endpoint status -w json | jq .
```

```json
[
  {
    "Endpoint": "127.0.0.1:2370",
    "Status": {
      "header": {
        "cluster_id": 16248273405584523000,
        "member_id": 3412179229575531000,
        "revision": 4,
        "raft_term": 2
      },
      "version": "3.4.18",
      "dbSize": 20480,
      "leader": 3412179229575531000,
      "raftIndex": 12,
      "raftTerm": 2,
      "raftAppliedIndex": 12,
      "dbSizeInUse": 16384
    }
  },
  {
    "Endpoint": "127.0.0.1:2371",
    "Status": {
      "header": {
        "cluster_id": 16248273405584523000,
        "member_id": 1514320713169590800,
        "revision": 4,
        "raft_term": 2
      },
      "version": "3.4.18",
      "dbSize": 20480,
      "leader": 3412179229575531000,
      "raftIndex": 12,
      "raftTerm": 2,
      "raftAppliedIndex": 12,
      "dbSizeInUse": 16384
    }
  },
  {
    "Endpoint": "127.0.0.1:2372",
    "Status": {
      "header": {
        "cluster_id": 16248273405584523000,
        "member_id": 8920919569373610000,
        "revision": 4,
        "raft_term": 2
      },
      "version": "3.4.18",
      "dbSize": 20480,
      "leader": 3412179229575531000,
      "raftIndex": 12,
      "raftTerm": 2,
      "raftAppliedIndex": 12,
      "dbSizeInUse": 16384
    }
  }
]
```

as you can see, raftIndex (currently at 12) is distinct from store version (currently at 4)

liggitt replied:
I think the test exercises the behavior we want to check, but the comment is implying things that aren't really accurate (that the list resourceVersion depends on the piggy-backing behavior). simplify the comment to "list resourceVersion only changes when writes are performed, not reads"

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/linearized_read_test.go:
```diff
@@ -0,0 +1,64 @@
+	}
+	originalRevision := out.ResourceVersion
+
+	for i := 0; i < 100; i++ {
```
no need for 100 reads... even just 1 or 2 proves the point

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go:
```diff
@@ -382,14 +387,41 @@ func testCheckEventType(t *testing.T, expectEventType watch.EventType, w watch.I
 	}
 }
 
+func resourceVersionNoLaterThan(sentinel string) func(string) error {
```
am I misreading this, or is this actually returning a function that enforces "not earlier than $sentinel"

stevekuznetsov replied:
This is imprecise language. I will update it to use older/earlier

liggitt replied:
should this be named `resourceVersionNoOlderThan` instead?

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go:
```diff
@@ -353,8 +353,13 @@ func TestProgressNotify(t *testing.T) {
 	}
-	result := &example.Pod{ObjectMeta: metav1.ObjectMeta{ResourceVersion: out.ResourceVersion}}
-	testCheckResult(t, watch.Bookmark, w, result)
+	testCheckResultFunc(t, watch.Bookmark, w, func(object runtime.Object) error {
```
this change is weakening the check, since it is no longer verifying it gets an object containing only the resource version

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go:
```diff
@@ -353,8 +353,13 @@ func TestProgressNotify(t *testing.T) {
+		if !ok {
+			return fmt.Errorf("got %T, not *example.Pod", object)
+		}
+		return resourceVersionNoLaterThan(out.ResourceVersion)(pod.ResourceVersion)
```
(as an aside, I find the updated version harder to understand... not sure if there's a clearer way to express this)

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:
```diff
@@ -2124,8 +2131,8 @@ func TestListInconsistentContinuation(t *testing.T) {
 		t.Fatalf("Unexpected continuation token set")
 	}
 	expectNoDiff(t, "incorrect third page", []example.Pod{*preset[2].storedObj}, out.Items)
-	if out.ResourceVersion != lastRVString {
```
`continueFromThirdItem` encodes a specific rv, so this should exactly match the `ResourceVersion` in the `inconsistentContinueFromSecondItem` list, right? Can we check against that and keep this check exact?

liggitt commented:
/lgtm
/approve
/milestone v1.25
